### DISCUSSION TITLE
Balance and hardlinks?

### DISCUSSION THREAD

**Discussion by FlorentLM** [trapexit/mergerfs#1385] [Q&A]
Hi,

I have many hardlinked files in the following structure:

```
A              +            B
/disk1                      /disk2 
|                           | 
+-- /dir1                  +-- /dir3 
|   |                       |   | 
|   +-- file1_orig          |   +-- file2_orig
|                           | 
+-- /dir2                   +-- /dir4 
|   |                       |   | 
|   +-- file1_hardlink      |   +-- file2_hardlink 
```

What if I add a third drive and call mergerfs.balance? Will the files and their corresponding hardlink be moved together? Or will this break the hardlinks?

Thanks :)

**Comment by trapexit** [comment]
The balance tool is an extremely simple thing it just moves files around. It does not consider links, symlinks, etc.

**Comment by FlorentLM** [comment]
Oh okay, thanks for lightning fast reply!
So just to be clear, there is currently no way to balance a pool while being hardlink-aware, right?

Might write something myelf and PR it, if that's of any use! :)

**Reply by trapexit** [reply]
No. The balance tool was never intended to be a serious tool. I never intended to treat mergerfs like DrivePool where it always is rebalancing. 

My preference is to build a new tool or feature into mergerfs all together but it is a lower priority to other work.

**Comment by eatnumber1** [comment]
I've been working on a very similar problem. I'm not interested in balancing (in fact, I want to be extremely unbalanced to avoid drive spinups), but I do want to move a directory _and all its hardlinks_ from one branch to another.

This is the hard part needed to make it work: https://github.com/eatnumber1/find-with-hardlinks/blob/master/find-with-hardlinks. Using that, you can get a list of all the files in a directory *and all the hardlinks to those files no matter where they are in the filesystem*. Then using an rsync like the following, you can move it:

```shell
$ env -C /mnt/space find-with-hardlinks --print0 --inode_sort --search_dir /mnt/space -- -mtime +30 -ctime +30 -Btime +30 -- Plex > files.txt
$ env -C /mnt/space rsync --archive --acls --xattrs --hard-links --inplace '--info=progress2' --preallocate --whole-file --no-compress --sparse --specials --devices --one-file-system --from0 '--files-from=files.txt' --remove-source-files /mnt/space /mnt/archive
```

That leaves behind some empty directories, which you can delete with

```shell
$ env -C /mnt/space find -O3 Plex -xdev -type d -empty -delete
```
